[docs] Add themeable component guide (#37908)
mui/material-ui: docs/data/joy/guides/themeable-component/themeable-component.md

# Themeable Component

<p class="description">Create your own themeable component with Joy UI theming feature.</p>

## Introduction

Joy UI provides a powerful theming feature that lets you to add your own component to the theme as if it's one of the built-in components.

If you are building a component library on top of Joy UI, you can follow the step-by-step guide below on how to create a custom component that is themeable across multiple projects.

Otherwise, you can skip the step-by-step guide and use the [template](#template) as a starting point for your component.

:::info
You don't need to connect your component to the theme if you are only using it in a single project.
:::

## Step-by-step guide

We are going to build a statistics component that looks like this:

{{"demo": "StatComponent.js", "hideToolbar": true}}

### 1. Create the component slots

> The slots let your users customize each part of the component by targeting the slot's name in [theme's styleOverrides](/joy-ui/customization/themed-components/#theme-style-overrides).

Author: Yes, that sounds better.

+   const { value, unit, variant, ...other } = props;
+
+   const ownerState = { ...props, variant };

    return (
-      <StatRoot ref={ref} {...other}>
-        <StatValue>{value}</StatValue>
-        <StatUnit>{unit}</StatUnit>
-      </StatRoot>
+      <StatRoot ref={ref} ownerState={ownerState} {...other}>
+        <StatValue ownerState={ownerState}>{value}</StatValue>
+        <StatUnit ownerState={ownerState}>{unit}</StatUnit>
+      </StatRoot>
    );
  });
```

Then you can read `ownerState` in the slot to style it based on the `variant` prop.

```diff
  const StatRoot = styled('div', {
    name: 'JoyStat',
    slot: 'root',
-  })(({ theme }) => ({
+  })(({ theme, ownerState }) => ({
    display: 'flex',
    flexDirection: 'column',
    gap: theme.spacing(0.5),
    padding: theme.spacing(3, 4),
    backgroundColor: theme.palette.background.paper,
    borderRadius: theme.shape.borderRadius,
    boxShadow: theme.shadows[2],
    letterSpacing: '-0.025em',
    fontWeight: 600,
+   ...ownerState.variant === 'outlined' && {
+    border: `2px solid ${theme.palette.divider}`,
+   },
  }));
```

### 4. Support theme default props

> To let your users customize the default props of your component for different projects, you need to use the `useThemeProps` API.

Reviewer: ```suggestion
To customize your component's default props for different projects, you need to use the `useThemeProps` API.
```

```diff
+ import { useThemeProps } from '@mui/joy/styles';

- const Stat = React.forwardRef(function Stat(props, ref) {
+ const Stat = React.forwardRef(function Stat(inProps, ref) {
+   const props = useThemeProps({ props: inProps, name: 'JoyStat' });
    const { value, unit, ...other } = props;

    return (
      <StatRoot ref={ref} {...other}>
        <StatValue>{value}</StatValue>
        <StatUnit>{unit}</StatUnit>
      </StatRoot>
    );
  });
```

> Then your users can customize the default props of your component like this:

Reviewer: ```suggestion
Then you can customize the default props of your component like this:
```

```js
import { extendTheme } from '@mui/joy/styles';

const theme = extendTheme({
  components: {
    JoyStat: {
      defaultProps: {
        variant: 'outlined',
      },
    },
  },
});
```

## TypeScript

> If you use TypeScript, you should create interfaces for the component props and ownerState.

Reviewer: ```suggestion
If you use TypeScript, you must create interfaces for the component props and ownerState:
```

  // but don't want to expose to the users
}
```

Then you can use them in the component and slots.

```js
const StatRoot = styled('div', {
  name: 'JoyStat',
  slot: 'root',
})<{ ownerState: StatOwnerState }>(({ theme, ownerState }) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(0.5),
  padding: theme.spacing(3, 4),
  backgroundColor: theme.palette.background.paper,
  borderRadius: theme.shape.borderRadius,
  boxShadow: theme.shadows[2],
  letterSpacing: '-0.025em',
  fontWeight: 600,
  // typed-safe access to the `variant` prop
  ...(ownerState.variant === 'outlined' && {
    border: `2px solid ${theme.palette.divider}`,
    boxShadow: 'none',
  }),
}));

// …do the same for other slots

const Stat = React.forwardRef<HTMLDivElement, StatProps>(function Stat(inProps, ref) {
  const props = useThemeProps({ props: inProps, name: 'JoyStat' });
  const { value, unit, variant, ...other } = props;

  const ownerState = { ...props, variant };

  return (
    <StatRoot ref={ref} ownerState={ownerState} {...other}>
      <StatValue ownerState={ownerState}>{value}</StatValue>
      <StatUnit ownerState={ownerState}>{unit}</StatUnit>
    </StatRoot>
  );
});
```

> Finally, you have to add the Stat component the theme types.

Reviewer: ```suggestion
Finally, add the Stat component to the theme types.
```

```ts
import { Theme, StyleOverrides } from '@mui/joy/styles';
import { StatProps, StatOwnerState } from '/path/to/Stat';

declare module '@mui/joy/styles' {
  interface Components {
    JoyStat?: {
      defaultProps?: Partial<StatProps>;
      styleOverrides?: StyleOverrides<StatProps, StatOwnerState, Theme>;
    };
  }
}
```

:::success

> 🍾 Congratulations! You have successfully built a themeable component!.

Reviewer: I would cut this (or at least remove the period)

```suggestion
```

:::

---

## Template

> You can use the following template to create your own component. It is the full code of the `Stat` component in this guide.

Reviewer: Reiterating the purpose of the template for anyone who skipped straight to it and might not be sure what they're looking at.

```suggestion
This template is the final product of the step-by-step guide above, demonstrating how to build a custom component that can be styled with the theme as if it was a built-in component.
```
